### PR TITLE
[pose control] Header PoseStamped consideration added

### PR DESCRIPTION
**Pull request by Fireronin**
Based on https://github.com/RobotecAI/robotec-o3de-tools/issues/22

### CODE REVIEW

File: Gems/ROS2PoseControl/Code/Source/Clients/ROS2PoseControl.cpp [hunk 1]
```diff
@@ -152,18 +152,55 @@ namespace ROS2PoseControl
...
+                else
+                {
+                    geometry_msgs::msg::TransformStamped transformStamped;
+                    if (m_tf_buffer->canTransform(
+                        m_configuration.m_referenceFrame.c_str(),
+                        msg->header.frame_id.c_str(),
+                        tf2::TimePointZero))
+                    {
+                        transformStamped = m_tf_buffer->lookupTransform(
+                            m_configuration.m_referenceFrame.c_str(),
```
[the remark this answers is not in the document]
Comment (Fireronin) [reply Gems/ROS2PoseControl/Code/Source/Clients/ROS2PoseControl.cpp:184]: Oh, but I think that should do the trick anyway the same code is used bellow in the GetCurrentTransformViaTF2 function

[the remark this answers is not in the document]
Comment (Fireronin) [reply Gems/ROS2PoseControl/Code/Source/Clients/ROS2PoseControl.cpp:184]: https://github.com/RobotecAI/robotec-o3de-tools/issues/36

**Review by michalpelka — APPROVED** [review]

File: Gems/ROS2PoseControl/Code/Source/Clients/ROS2PoseControl.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (adamdbrw) [reply Gems/ROS2PoseControl/Code/Source/Clients/ROS2PoseControl.cpp:184]: +1 for exception handling

**Review by arturkamieniecki — APPROVED** [review]

File: Gems/ROS2PoseControl/Code/Source/Clients/ROS2PoseControl.cpp [hunk 2]
```diff
@@ -90,32 +90,44 @@ namespace ROS2PoseControl
...
+                    tf2::TimePointZero);
+                m_tf2WarningShown = false;
+            } catch (const tf2::TransformException& ex)
+            {
+                errorString = ex.what();
+                exceptionThrown = true;
+            }
         }
-        else
+        if (exceptionThrown || !errorString.empty())
```
[the remark this answers is not in the document]
Comment (Fireronin) [reply Gems/ROS2PoseControl/Code/Source/Clients/ROS2PoseControl.cpp:121]: Both canTransform and lookupTransform can produce error so I wanted to handle both

[the remark this answers is not in the document]
Comment (arturkamieniecki) [reply Gems/ROS2PoseControl/Code/Source/Clients/ROS2PoseControl.cpp:121]: Resolved after offline discussion.

File: Gems/ROS2PoseControl/Code/Source/Clients/ROS2PoseControl.cpp [hunk 3]
```diff
@@ -90,32 +90,44 @@ namespace ROS2PoseControl
...
+            {
+                transformStamped = m_tf_buffer->lookupTransform(
+                    targetFrameCStr,
+                    sourceFrameCStr,
+                    tf2::TimePointZero);
+                m_tf2WarningShown = false;
+            } catch (const tf2::TransformException& ex)
+            {
+                errorString = ex.what();
+                exceptionThrown = true;
```
Comment (michalpelka) [inline Gems/ROS2PoseControl/Code/Source/Clients/ROS2PoseControl.cpp:118]: Why doesn't you return here with ex.what() as fail?

File: Gems/ROS2PoseControl/Code/Source/Clients/ROS2PoseControl.cpp [hunk 4]
```diff
@@ -90,32 +90,44 @@ namespace ROS2PoseControl
         }
     }
 
-    AZ::Outcome<AZ::Transform, const char*> ROS2PoseControl::GetCurrentTransformViaTF2()
+    AZ::Outcome<AZ::Transform, const char*> ROS2PoseControl::GetCurrentTransformViaTF2(
```
Comment (michalpelka) [inline Gems/ROS2PoseControl/Code/Source/Clients/ROS2PoseControl.cpp:93]: I am not big fan of const char* in this context. 
Why? I do not know who is owner of this pointer and lifetime. 
As API user I do know if it is safe to use, when it will be freed, if should I free it by myself. Please adjust.

File: Gems/ROS2PoseControl/Code/Source/Clients/ROS2PoseControl.cpp [hunk 5]
```diff
@@ -43,10 +43,10 @@ namespace ROS2PoseControl
     void ROS2PoseControl::Activate()
     {
         auto ros2Node = ROS2::ROS2Interface::Get()->GetNode();
```
Comment (michalpelka) [inline Gems/ROS2PoseControl/Code/Source/Clients/ROS2PoseControl.cpp:45]: Assert on ros2 node.

**Review by michalpelka — CHANGES_REQUESTED** [review]

File: Gems/ROS2PoseControl/Code/Source/Clients/ROS2PoseControl.cpp (hunk 3 above)
Comment (Fireronin) [reply Gems/ROS2PoseControl/Code/Source/Clients/ROS2PoseControl.cpp:118]: I wanted to have neat debug print

**Review by michalpelka — APPROVED** [review]

**Review by michalpelka — CHANGES_REQUESTED** [review]
I cannot accept - I set it to move in base_link with identity transform - this should give no movement, however it caused robot movement.

**Comment by Fireronin** [comment]
Rebased and tested

**Comment by pijaro** [comment]
I can't make it work in my example.

I put two entities on the scene with ROS 2 frames (`base_link` and `target`):

image
image

I've added the `ROS2PoseControl` to `base_link` model, set tracking to true, and published the message:

```
ros2 topic pub /goal_pose geometry_msgs/msg/PoseStamped "{header:{frame_id: target}, pose:{position:{x: 1.0}}}" -1
```

It was expected that `base_link` model would teleport to 1m in front of `target`, but I got the warning:
```
[Warning] (ROS2PoseControl) - No entity with tag found .
[Warning] (ROS2PositionControl) - Could not transform map to target, error: Invalid frame ID "map" passed to canTransform argument target_frame - frame does not exist
```
and nothing happened.

It works when I set the frame to `base_link`, then it will jump 1m on X axis:
```
ros2 topic pub /goal_pose geometry_msgs/msg/PoseStamped "{header:{frame_id: base_link}, pose:{position:{x: 1.0}}}" -1
```

**Comment by Fireronin** [comment]
Generally, I have written a more detailed documentation https://github.com/RobotecAI/robotec-o3de-tools/blob/kr/pose-controll-better-documentation/doc/UserGuide/PoseControl.md

It will be merged in another pr, but first this needs to be resolved, you are missing transform from map frame to target frame. Therefore If you are saying I want to move this object in front of target frame, component tries to find location in relation to map.

This is because this component is generally made to work in ROS 2 tf positions

**Comment by Fireronin** [comment]
I see, in my testing I didn't encountered this as correct values were set when I was using goal position and then they persisted even when they were hidden, now I made thise settings visible once more

**Comment by pijaro** [comment]
TF tree in my example is a valid ROS 2 TF tree and the proposed system should also work. 

There is no reason to require `map` frame to exist.

**Comment by pijaro** [comment]
As for the reference frame, I don't think this should be a requirement as well.

We have TF API and we can get any transformation in a tree we want. Why do we need to stick to one, specified reference frame? We can get a transform `base_link` -> `target` and then use that to move `base_link` the way we want.

**Comment by Fireronin** [comment]
Tldr, this will introduce even more feature creep, currently there are like 4 cases for goal messages now it will be 5

**Comment by pijaro** [comment]
I'm not sure what cases you are referring to exactly. 

Let's make it work for all TF trees and with only the essential configuration needed. 

If the code is starting to look bloated, then maybe some alternative approaches should be considered.

**Comment by pijaro** [comment]
> Generally, I have written a more detailed documentation https://github.com/RobotecAI/robotec-o3de-tools/blob/kr/pose-controll-better-documentation/doc/UserGuide/PoseControl.md

Is there a reason not to include the readme update here, in this PR? Why splitting it?

**Comment by Fireronin** [comment]
@michalpelka Can you take a look one more time ?

**Comment by arturkamieniecki** [comment]
I'm closing this PR due to a newer one: https://github.com/RobotecAI/robotec-o3de-tools/pull/88.
